### ISSUE TITLE
Tests with database access crash starcounter service

### ISSUE THREAD

**Issue by gsimion** [Starcounter/Home#410]
Version: 2.4.0.6025.
I am experiencing problems when running the tests with database access while the starcounter service is running as local service. The tests have been written following the [documentation ](https://docs.starcounter.io/v/2.4/how-to-guides/how-to-write-tests-with-database-access) available translated to a NUnit context. 
```
// Start server to interact with host 
// Stop host in case a database is already running
// Recreate the database to start clean
_cli = new StarAdminCli(Properties.Settings.Default.TestDbInstance);
_cli.StartServer()
                .StopHost()
                .RecreateDatabase(<PORT>);

// Load the database into the app host and
// point to the assembly to test
_host = new AppHostBuilder()
                .UseDatabase(<NAME>)
                .UseApplication(typeof(<APP>).Assembly)
                .Build();
            
_host.Start();
```
I am running with full administrator credentials. When running tests without lunching the service first (i.e. service.exe, scnetworkgateway.exe, etc running as Admin) the tests go through without any problem.

**Procedure to reproduce**:

1. Services -> Starcounter Server Service -> start
2. Run tests with admin credentials

**Expected behavior**: 

- Test execution

**Actual behavior**: 

- New scnetworkgateway.exe process created running as local service
- The newly created process collides the existing one, resulting in a full Starcounter service crash
- Exception occurs in `_host.Start()`

Exception message:
Message: OneTimeSetUp: System.AggregateException : One or more errors occurred.
  ----> System.Net.Http.HttpRequestException : An error occurred while sending the request.
  ----> System.Net.WebException : The underlying connection was closed: An unexpected error occurred on a receive.
  ----> System.IO.IOException : Unable to read data from the transport connection: An existing connection was forcibly closed by the remote host.
  ----> System.Net.Sockets.SocketException : An existing connection was forcibly closed by the remote host

Event log:
Faulting application name: scnetworkgateway.exe, version: 0.0.0.0, time stamp: 0x5ae70818
Faulting module name: ucrtbase.dll, version: 10.0.16299.248, time stamp: 0xe71e5dfe
Exception code: 0xc0000409
Faulting application path: C:\Program Files\Starcounter\scnetworkgateway.exe
Faulting module path: C:\Windows\System32\ucrtbase.dll
The Starcounter Server Service service terminated unexpectedly.

**Comment by miyconst** [comment]
cc @per-samuelsson

**Comment by per-samuelsson** [comment]
Assuming this `StarAdminCli` class is similar in semantics to how `TestSetup` is sampled in referenced documentation, I can't see any obvious problem in the code.

To fully find out what causes this, we need to setup an environment and step through to see where the inconsistency occur. How urgent is it? And @miyconst, do you want me to do that? It's probably a pretty general thing relating to process management.

**Comment by miyconst** [comment]
@per-samuelsson this is quite important, but I will first setup the test environment and try to reproduce it.

**Comment by miyconst** [comment]
I have reproduced the issue and have a clean VM where it can be investigated. Will share the credentials on Slack.

**Comment by per-samuelsson** [comment]
Not having any great success with this, but I start to suspect it's not necessarily related just to testing.

Looking at last section in the OP:

> Event log:
Faulting application name: scnetworkgateway.exe, version: 0.0.0.0, time stamp: 0x5ae70818
Faulting module name: ucrtbase.dll, version: 10.0.16299.248, time stamp: 0xe71e5dfe
Exception code: 0xc0000409
Faulting application path: C:\Program Files\Starcounter\scnetworkgateway.exe
Faulting module path: C:\Windows\System32\ucrtbase.dll
The Starcounter Server Service service terminated unexpectedly.

it closely resemble https://github.com/Starcounter/level1/issues/4456

Also, when I build latest sources of 'develop' and do a clean install (from L1 output folder) and configure **Production** in the installer (meaning that the service should run), the *service fail to start* due to problem starting the gateway. In Windows logs, I see this

> Service cannot be started. System.Exception: ScErrStartNetworkGateway (SCERR2110): Problems starting network gateway process.
Version: 0.0.0.
Help page: https://docs.starcounter.io/?q=SCERR2110.
   at Starcounter.Server.ServerServiceProcess.WaitUntilServerIsOnlineOrSignalExit[T](Func`2 checkExited, T context) in 
\Level1\src\Server\Starcounter.Server\ServerServiceProcess.cs:line 220
   at Starcounter.Server.ServerServiceProcess.OnStart(String[] args) in \Level1\src\Server\Starcounter.Server\ServerServiceProcess.cs:line 150
   at System.ServiceProcess.ServiceBase.ServiceQueuedMainCallback(Object state)

but also this in Windows System Log section:

> Application popup: Microsoft Visual C++ Runtime Library : Debug Error!
> Program: ...\Level1\bin\Debug\scnetworkgateway.exe
> abort() has been called
> (Press Retry to debug the application)

@alemoi, anything you know about? Can you successfully start the service/gw with latest bits?

**Comment by alemoi** [comment]
> Also, when I build latest sources of 'develop' and do a clean install (from L1 output folder) and configure Production in the installer (meaning that the service should run), the service fail to start due to problem starting the gateway.

> @alemoi, anything you know about? Can you successfully start the service/gw with latest bits?

Tried the same - no failure, the service and Admin runs properly.

Nothing in logs you have?

**Comment by alemoi** [comment]
I have investigated the issue and it seems related to different privileges, since Starcounter is running as a service.
Isn't it a known issue related to shared memory segments?
![image](https://user-images.githubusercontent.com/208134/39863987-96262bd2-5448-11e8-8355-63fdec01a1e4.png)
Basically `scnetworkgateway` can't attach to shared memory segment made by `sccode`, maybe because its running under different privileges? There are no changes in `scnetworkgateway` in that respect. Maybe some changes in `sccode` @bigwad ?

**Comment by per-samuelsson** [comment]
> I wonder what happens when I run the tests, in terms of processes. Are processes created in user mode?

Well, in this particular story, I think it's like this:

1. All servicing processes (e.g. gateway, shared host, scdata) will be running in session 0.
2. The testing process (and "self-host") is running in user mode.

In fact, we never tested such case I think. I imagined that if you configure Starcounter to install in production mode, that means you would not run unit testing.

**Comment by alemoi** [comment]
After fixing the shared memory problem, there is another one appeared, seems totally unrelated: https://github.com/Starcounter/level1/pull/4638#issuecomment-390224418
CC @per-samuelsson if you know anything about it.

**Comment by per-samuelsson** [comment]
> CC @per-samuelsson if you know anything about it.

See comment in PR.

**Comment by miyconst** [comment]
Closing as won't fix in Starcounter 2.4 and not applicable for Starcounter 3.0.
